### PR TITLE
Add ability to flag HLE functions to clear a bit of stack.

### LINKED ISSUE #6374: Sengoku Hime 4 black screen
It's not hanging,we can hear sound.A lot of files can not be found in the log.the video was not black before,video is black since https://github.com/unknownbrackets/ppsspp/commit/2042672ec718b177268f590c9f10e43a6fbf6a00

info log:
https://gist.github.com/daniel229/22561bdfcc55fa8c603d

debug log(rename jpg to rar):
ppsspplog

### PR DESCRIPTION
**Pull request by hrydgard**
Idea from JPCSP, pointed out by sum2012.

Appears to fix #6374.

### CODE REVIEW

**Comment by unknownbrackets** [comment]
Perhaps I'm testing wrong, but I'm using this to measure stack usage by these funcs:

```c++
static inline u32 *getSP() {
	u32 *sp;
    asm("move %0, $sp" : "=r" (sp));
    return sp;
}

int usage_end;
const char *usage_title;
u32 *usage_sp;
int usage_i;

void testUsageStart() {
	for (usage_i = -0x200; usage_i < 0; ++usage_i) {
		usage_sp[usage_i] = 0x1337C0DE;
	}
}

void testUsageEnd() {
	usage_end = -1;
	for (usage_i = -0x200; usage_i < 0; ++usage_i) {
		if (usage_sp[usage_i] != 0x1337C0DE) {
			usage_end = usage_i;
			break;
		}
	}

	checkpoint("  %s: %d bytes", usage_title, -usage_end * 4);
	for (usage_i = usage_end; usage_i < 0; ++usage_i) {
		schedf("    %08x", usage_sp[usage_i]);
	}
	schedf("\n");
}

#define TEST_USAGE(expr) \
	usage_sp = getSP(); \
	usage_title = #expr; \
	testUsageStart(); \
	expr; \
	testUsageEnd();
```

Notably, even `sceKernelDelayThread()` lists 4 bytes, so I guess I'm leaking something there:
```
[r]   sceKernelDelayThread(10): 4 bytes
    deadbeef
```

Anyway, the sceFont functions give me these values at least for the lib version I'm using:
```
[x] Init
[r]   FontLibraryHandle fontLib = sceFontNewLib(&libParams, &error): 1440 bytes
[x]   sceFontSetResolution(fontLib, 1.0f, 1.0f): 4 bytes
[x]   sceFontGetNumFontList(fontLib, &error): 4 bytes
[r]   sceFontGetFontList(fontLib, styles, 4): 796 bytes
[x]   sceFontFindOptimumFont(fontLib, styles, &error): 240 bytes
[x]   sceFontFindFont(fontLib, styles, &error): 64 bytes
[x]   sceFontGetFontInfoByIndexNumber(fontLib, styles, 0): 32 bytes
[r]   FontHandle f = sceFontOpen(fontLib, 0, 0, &error): 1120 bytes
[x] RES=088399b0, 00000000
[x]   int res = sceFontGetFontInfo(f, &info): 4 bytes
[r]   sceFontGetCharInfo(f, 'A', &charInfo): 272 bytes
[x]   sceFontGetShadowInfo(f, 'A', &charInfo): 336 bytes
[x]   sceFontGetCharImageRect(f, 'A', &rect): 288 bytes
[x]   sceFontGetShadowImageRect(f, 'A', &rect): 336 bytes
[x]   sceFontGetCharGlyphImage(f, 'A', &im): 288 bytes
[x]   sceFontGetCharGlyphImage_Clip(f, 'A', &im, 0, 0, 1, 1): 304 bytes
[x]   sceFontGetShadowGlyphImage(f, 'A', &im): 352 bytes
[x]   sceFontGetShadowGlyphImage_Clip(f, 'A', &im, 0, 0, 1, 1): 368 bytes
[x]   sceFontClose(f): 84 bytes
[x]   sceFontPixelToPointH(fontLib, 1.0f, &error): 16 bytes
[x]   sceFontPixelToPointV(fontLib, 1.0f, &error): 16 bytes
[x]   sceFontPointToPixelH(fontLib, 1.0f, &error): 4 bytes
[x]   sceFontPointToPixelV(fontLib, 1.0f, &error): 4 bytes
[x]   sceFontSetAltCharacterCode(fontLib, 'Z'): 4 bytes
[x]   sceFontDoneLib(fontLib): 44 bytes
```

Not sure if it's necessary to replicate all of them, though...

-[Unknown]

**Comment by daniel229** [comment]
Replace all the value from @unknownbrackets ,the game is still working.Missing this one sceFontOpenUserMemory that JPCSP has 0x440 bytes stack.
```
const HLEFunction sceLibFont[] = {
	{0x67f17ed7, WrapU_UU<sceFontNewLib>, "sceFontNewLib", HLE_CLEAR_STACK_BYTES, 0x5A0},
	{ 0x574b6fbc, WrapI_U<sceFontDoneLib>, "sceFontDoneLib", HLE_CLEAR_STACK_BYTES, 0x2C },
	{0x48293280, WrapI_UFF<sceFontSetResolution>, "sceFontSetResolution"},
	{ 0x27f6e642, WrapI_UU<sceFontGetNumFontList>, "sceFontGetNumFontList"},
	{ 0xbc75d85b, WrapI_UUI<sceFontGetFontList>, "sceFontGetFontList", HLE_CLEAR_STACK_BYTES, 0x31C },
	{ 0x099ef33c, WrapI_UUU<sceFontFindOptimumFont>, "sceFontFindOptimumFont", HLE_CLEAR_STACK_BYTES, 0xF0 },
	{ 0x681e61a7, WrapI_UUU<sceFontFindFont>, "sceFontFindFont", HLE_CLEAR_STACK_BYTES, 0x40 },
	{0x2f67356a, WrapI_V<sceFontCalcMemorySize>, "sceFontCalcMemorySize"},
	{ 0x5333322d, WrapI_UUU<sceFontGetFontInfoByIndexNumber>, "sceFontGetFontInfoByIndexNumber", HLE_CLEAR_STACK_BYTES, 0x20 },
	{ 0xa834319d, WrapU_UUUU<sceFontOpen>, "sceFontOpen", HLE_CLEAR_STACK_BYTES, 0x460 },
	{0x57fcb733, WrapU_UCUU<sceFontOpenUserFile>, "sceFontOpenUserFile"},
	{0xbb8e7fe6, WrapU_UUUU<sceFontOpenUserMemory>, "sceFontOpenUserMemory"/*, HLE_CLEAR_STACK_BYTES, 0x440 from JPCSP*/},
	{ 0x3aea8cb6, WrapI_U<sceFontClose>, "sceFontClose", HLE_CLEAR_STACK_BYTES, 0x54 },
	{0x0da7535e, WrapI_UU<sceFontGetFontInfo>, "sceFontGetFontInfo"},
	{0xdcc80c2f, WrapI_UUU<sceFontGetCharInfo>, "sceFontGetCharInfo", HLE_CLEAR_STACK_BYTES, 0x110},
	{ 0xaa3de7b5, WrapI_UUU<sceFontGetShadowInfo>, "sceFontGetShadowInfo", HLE_CLEAR_STACK_BYTES, 0x150 },
	{ 0x5c3e4a9e, WrapI_UUU<sceFontGetCharImageRect>, "sceFontGetCharImageRect", HLE_CLEAR_STACK_BYTES, 0x120 },
	{ 0x48b06520, WrapI_UUU<sceFontGetShadowImageRect>, "sceFontGetShadowImageRect", HLE_CLEAR_STACK_BYTES, 0x150 },
	{0x980f4895, WrapI_UUU<sceFontGetCharGlyphImage>, "sceFontGetCharGlyphImage", HLE_CLEAR_STACK_BYTES, 0x120},
	{ 0xca1e6945, WrapI_UUUIIII<sceFontGetCharGlyphImage_Clip>, "sceFontGetCharGlyphImage_Clip", HLE_CLEAR_STACK_BYTES, 0x130 },
	{ 0x74b21701, WrapF_IFU<sceFontPixelToPointH>, "sceFontPixelToPointH", HLE_CLEAR_STACK_BYTES, 0x10 },
	{ 0xf8f0752e, WrapF_IFU<sceFontPixelToPointV>, "sceFontPixelToPointV", HLE_CLEAR_STACK_BYTES, 0x10 },
	{0x472694cd, WrapF_IFU<sceFontPointToPixelH>, "sceFontPointToPixelH"},
	{0x3c4b7e82, WrapF_IFU<sceFontPointToPixelV>, "sceFontPointToPixelV"},
	{0xee232411, WrapI_UU<sceFontSetAltCharacterCode>, "sceFontSetAltCharacterCode"},
	{ 0x568be516, WrapI_UUU<sceFontGetShadowGlyphImage>, "sceFontGetShadowGlyphImage", HLE_CLEAR_STACK_BYTES, 0x160 },
	{ 0x5dcf6858, WrapI_UUUIIII<sceFontGetShadowGlyphImage_Clip>, "sceFontGetShadowGlyphImage_Clip", HLE_CLEAR_STACK_BYTES, 0x170 },
	{0x02d7f94b, WrapI_U<sceFontFlush>, "sceFontFlush"},
};
```

**Comment by unknownbrackets** [comment]
Yeah I didn't do user file or memory yet.

-[Unknown]

**Comment by sum2012** [comment]
I agree that change this pull request 's value to @unknownbrackets 's

**Comment by hrydgard** [comment]
I'll change the values and merge tonight.

File: Core/HLE/sceFont.cpp [hunk 1]
```diff
@@ -1354,35 +1354,37 @@ static int sceFontGetShadowGlyphImage_Clip(u32 fontHandle, u32 charCode, u32 gly
...
+	{ 0x48293280, WrapI_UFF<sceFontSetResolution>, "sceFontSetResolution" },
+	{ 0x27f6e642, WrapI_UU<sceFontGetNumFontList>, "sceFontGetNumFontList" },
+	{ 0xbc75d85b, WrapI_UUI<sceFontGetFontList>, "sceFontGetFontList", HLE_CLEAR_STACK_BYTES, 0x31C },
+	{ 0x099ef33c, WrapI_UUU<sceFontFindOptimumFont>, "sceFontFindOptimumFont", HLE_CLEAR_STACK_BYTES, 0xF0 },
+	{ 0x681e61a7, WrapI_UUU<sceFontFindFont>, "sceFontFindFont", HLE_CLEAR_STACK_BYTES, 0x40 },
+	{ 0x2f67356a, WrapI_V<sceFontCalcMemorySize>, "sceFontCalcMemorySize" },
+	{ 0x5333322d, WrapI_UUU<sceFontGetFontInfoByIndexNumber>, "sceFontGetFontInfoByIndexNumber", HLE_CLEAR_STACK_BYTES, 0x20 },
+	{ 0xa834319d, WrapU_UUUU<sceFontOpen>, "sceFontOpen", HLE_CLEAR_STACK_BYTES, 0x460 },
+	{ 0x57fcb733, WrapU_UCUU<sceFontOpenUserFile>, "sceFontOpenUserFile" },
+	{ 0xbb8e7fe6, WrapU_UUUU<sceFontOpenUserMemory>, "sceFontOpenUserMemory"/*, HLE_CLEAR_STACK_BYTES, 0x440 from JPCSP*/ },
```
Comment (sum2012) [inline Core/HLE/sceFont.cpp]: I suggest use JPCSP value until @unknownbrackets test it.

**Comment by sum2012** [comment]
I am looking forward to @unknownbrackets test scepsmf\scepsmfplayer stack's value to improve video

File: Core/HLE/sceFont.cpp (hunk 1 above)
Comment (hrydgard) [inline Core/HLE/sceFont.cpp]: Done

Comment (sum2012) [inline Core/HLE/sceFont.cpp]: Thanks

**Pull request merged**